src/prelude/internals.mo

```diff
@@ -428,3 +428,153 @@ func @create_actor_helper(wasm_module_ : Blob, arg_ : Blob) : async Principal =
+var @timers : ?@Node = null;
+func @prune(n : ?@Node) : ?@Node = switch n {
+    case null null;
+    case (?n) {
+        if (n.expire[0] == 0) {
+            @prune(n.post) // by corollary
+        } else {
+            ?{ n with pre = @prune(n.pre); post = @prune(n.post) }
+        }
+    }
+};
+
+func @nextExpiration(n : ?@Node) : Nat64 = switch n {
+    case null 0;
+    case (?n) {
+        var exp = @nextExpiration(n.pre); // TODO: use the corollary for expire == 0
+        if (exp == 0) {
+            exp := n.expire[0];
+            if (exp == 0) {
+                exp := @nextExpiration(n.post)
+            }
+        };
+        exp
+    }
+};
+
+// Function called by backend to run eligible timed actions.
+// DO NOT RENAME without modifying compilation.
+func @timer_helper() : async () {
+    func Array_init<T>(len : Nat,  x : T) : [var T] {
+        (prim "Array.init" : <T>(Nat, T) -> [var T])<T>(len, x)
+    };
+    let now = (prim "time" : () -> Nat64)();
+    let exp = @nextExpiration @timers;
+    let prev = (prim "global_timer_set" : Nat64 -> Nat64) exp;
+
+    if (exp == 0) {
+        return
+    };
+
+    var gathered = 0;
+    let thunks = Array_init<?(() -> async ())>(10, null); // we want max 10
+
+    func gatherExpired(n : ?@Node) = switch n {
+        case null ();
+        case (?n) {
+            gatherExpired(n.pre);
+            if (n.expire[0] > 0 and n.expire[0] <= now and gathered < thunks.size()) {
+                thunks[gathered] := ?(n.job);
+                switch (n.delay) {
+                    case (?delay) {
+                        // re-add the node
+                        let expire = n.expire[0] + delay;
+                        // N.B. insert only works on pruned nodes
+                        func reinsert(m : ?@Node) : @Node = switch m {
+                            case null ({ n with expire = [var expire]; pre = null; post = null });
+                            case (?m) {
+                                assert m.expire[0] != 0;
+                                if (expire < m.expire[0]) ({ m with pre = ?reinsert(m.pre) })
+                                else ({ m with post = ?reinsert(m.post) })
+                            }
+                        };
+                        @timers := ?reinsert(@prune(@timers));
+                    };
+                    case _ ()
+                };
+                n.expire[0] := 0;
+                gathered += 1;
+            };
+            gatherExpired(n.post);
+        }
+    };
+
+    gatherExpired(@timers);
+
+    let futures = Array_init<?(async ())>(thunks.size(), null);
+    for (k in thunks.keys()) {
+        futures[k] := switch (thunks[k]) { case (?thunk) ?thunk(); case _ null };
+    };
+
+    for (f in futures.vals()) {
+        switch f { case (?f) try await f catch _ { }; case _ () }
+    };
+};
+
+var @lastId = 0;
+
+func @setTimer(delaySecs : Nat64, recurring : Bool, job : () -> async ()) : (id : Nat) {
```

Comment:
My previous comment was marked as outdated, but I guess the comment still stands: we better to align with Rust CDK and either implement in both languages `setTImer()` + `setTimerInterval()` or `setTimer(..., recurring)`...

Reply:
So this PR is not the last word about user-facing interfaces. This just adds the _primitives_ to a (basically) hidden internal module for the compiler. As such less is more. What we choose to export from the base library is a different issue and we may go with the Rust-like naming. But I suspect that this is a non-issue as I don't expect that both Rust and Motoko code will surface in the same project soon :-)